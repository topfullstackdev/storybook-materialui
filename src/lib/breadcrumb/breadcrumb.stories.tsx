import { Meta, Story } from '@storybook/react';

import { Breadcrumb, BreadcrumbProps } from './breadcrumb';

const items = [
  { text: 'HOME', href: '/' },
  { text: 'SUB-LINK', href: '/' },
  { text: 'CURRENT-PAGE' },
];

export default {
  component: Breadcrumb,
  title: 'Breadcrumb',
} as Meta;

const Template: Story<BreadcrumbProps> = function Template(args) {
  return <Breadcrumb {...args} />;
};

export const Primary = Template.bind({});
Primary.args = {
  items,
};
