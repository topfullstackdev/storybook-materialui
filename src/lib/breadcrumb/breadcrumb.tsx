import styled from '@emotion/styled';
import Breadcrumbs from '@mui/material/Breadcrumbs';
import { NavLink } from 'react-router-dom';

import { BrandColorNames, useBrandColor } from '@bai/utils/theme';

import { Text } from '../text/text';

export type BreadcrumbItem = {
  text: string;
  href?: string;
};

export interface BreadcrumbProps {
  items: BreadcrumbItem[];
}

const StyledBreadcrumbs = styled(Breadcrumbs)`
  font-size: 12px;
  & a {
    text-decoration-color: #1eabe3 !important;
  }
`;

/**
 * This component should generally not be used directly.
 * Instead, use @bai/features/use-breadcrumb's exports useBreadcrumb and ContextBreadcrumb.
 * This allows the breadcrumb to be rendered anywhere in the component tree,
 * while the props from the breadcrumb can also be provided from anywhere
 * in the component tree.
 */
export function Breadcrumb(props: BreadcrumbProps) {
  const { items } = props;
  const linkColor = useBrandColor(
    BrandColorNames.SECONDARY,
    (theme) => theme.palette.primary.main,
  );
  return (
    <StyledBreadcrumbs>
      {items.map((item) => {
        const { text, href } = item;
        const key = text + (href || '');
        if (href) {
          return (
            <NavLink data-cy={`breadcrumb-item:${text}`} to={href} key={key}>
              <Text
                underlined
                color={linkColor}
                variant="caption"
                component="span"
              >
                {text}
              </Text>
            </NavLink>
          );
        }
        return (
          <Text
            data-cy={`Current Page Breadcrumb:${text}`}
            color="text.primary"
            variant="caption"
            component="span"
            key={key}
          >
            {text}
          </Text>
        );
      })}
    </StyledBreadcrumbs>
  );
}

export default Breadcrumb;
