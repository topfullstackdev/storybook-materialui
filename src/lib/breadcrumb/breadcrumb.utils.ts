import { BreadcrumbItem } from './breadcrumb';

export interface BreadcrumbItemDefinition {
  text: string;
  href?: string;
  parent?: string;
}

export function getBreadcrumbItems(
  url: string,
  itemDefinitions: BreadcrumbItemDefinition[],
): BreadcrumbItem[] {
  return [...itemDefinitions]
    .reverse()
    .reduce((items, item) => {
      const lastItem = items[0];
      if (lastItem) {
        return item.text === lastItem.parent ? [item, ...items] : items;
      }
      return item.href && url.includes(item.href)
        ? [{ text: item.text, parent: item.parent }] // Leave off href on first match
        : [];
    }, [] as BreadcrumbItemDefinition[])
    .map(({ text, href }) => ({ text, href }));
}
