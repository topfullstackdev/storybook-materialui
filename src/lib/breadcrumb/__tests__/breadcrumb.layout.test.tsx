import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

import { Breadcrumb } from '../breadcrumb';

const items = [
  { text: 'HOME', href: '/' },
  { text: 'SUB-LINK', href: '/sub-link' },
  { text: 'CURRENT-PAGE' },
];

describe('Breadcrumb', () => {
  it('should render successfully', () => {
    const { container } = render(
      <BrowserRouter>
        <Breadcrumb items={items} />
      </BrowserRouter>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should be added the link', () => {
    const { container, getByText } = render(
      <BrowserRouter>
        <Breadcrumb items={items} />
      </BrowserRouter>,
    );
    expect(getByText('SUB-LINK').closest('a')).toHaveAttribute(
      'href',
      '/sub-link',
    );
    expect(container).toMatchSnapshot();
  });
});
