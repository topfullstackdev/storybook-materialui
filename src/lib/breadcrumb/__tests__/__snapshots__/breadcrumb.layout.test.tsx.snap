// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb should be added the link 1`] = `
.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.5;
  letter-spacing: 0.00938em;
  color: rgba(0, 0, 0, 0.6);
  font-size: 12px;
}

.emotion-1 a {
  text-decoration-color: #1eabe3!important;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0;
  margin: 0;
  list-style: none;
}

.emotion-4 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 0.75rem;
  line-height: 1.66;
  letter-spacing: 0.03333em;
  color: #1976d2;
  font-weight: normal;
  font-style: normal;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  font-family: Verdana,Arial,sans-serif;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  margin-left: 8px;
  margin-right: 8px;
}

.emotion-10 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 0.75rem;
  line-height: 1.66;
  letter-spacing: 0.03333em;
  color: rgba(0, 0, 0, 0.87);
  font-weight: normal;
  font-style: normal;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: Verdana,Arial,sans-serif;
}

<div>
  <nav
    class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root emotion-0 emotion-1"
  >
    <ol
      class="MuiBreadcrumbs-ol emotion-2"
    >
      <li
        class="MuiBreadcrumbs-li"
      >
        <a
          aria-current="page"
          class="active"
          data-cy="breadcrumb-item:HOME"
          href="/"
        >
          <span
            class="MuiTypography-root MuiTypography-caption emotion-3 emotion-4"
          >
            HOME
          </span>
        </a>
      </li>
      <li
        aria-hidden="true"
        class="MuiBreadcrumbs-separator emotion-5"
      >
        /
      </li>
      <li
        class="MuiBreadcrumbs-li"
      >
        <a
          class=""
          data-cy="breadcrumb-item:SUB-LINK"
          href="/sub-link"
        >
          <span
            class="MuiTypography-root MuiTypography-caption emotion-3 emotion-4"
          >
            SUB-LINK
          </span>
        </a>
      </li>
      <li
        aria-hidden="true"
        class="MuiBreadcrumbs-separator emotion-5"
      >
        /
      </li>
      <li
        class="MuiBreadcrumbs-li"
      >
        <span
          class="MuiTypography-root MuiTypography-caption emotion-3 emotion-10"
          data-cy="Current Page Breadcrumb:CURRENT-PAGE"
        >
          CURRENT-PAGE
        </span>
      </li>
    </ol>
  </nav>
</div>
`;

exports[`Breadcrumb should render successfully 1`] = `
.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.5;
  letter-spacing: 0.00938em;
  color: rgba(0, 0, 0, 0.6);
  font-size: 12px;
}

.emotion-1 a {
  text-decoration-color: #1eabe3!important;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0;
  margin: 0;
  list-style: none;
}

.emotion-4 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 0.75rem;
  line-height: 1.66;
  letter-spacing: 0.03333em;
  color: #1976d2;
  font-weight: normal;
  font-style: normal;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  font-family: Verdana,Arial,sans-serif;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  margin-left: 8px;
  margin-right: 8px;
}

.emotion-10 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 0.75rem;
  line-height: 1.66;
  letter-spacing: 0.03333em;
  color: rgba(0, 0, 0, 0.87);
  font-weight: normal;
  font-style: normal;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: Verdana,Arial,sans-serif;
}

<div>
  <nav
    class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root emotion-0 emotion-1"
  >
    <ol
      class="MuiBreadcrumbs-ol emotion-2"
    >
      <li
        class="MuiBreadcrumbs-li"
      >
        <a
          aria-current="page"
          class="active"
          data-cy="breadcrumb-item:HOME"
          href="/"
        >
          <span
            class="MuiTypography-root MuiTypography-caption emotion-3 emotion-4"
          >
            HOME
          </span>
        </a>
      </li>
      <li
        aria-hidden="true"
        class="MuiBreadcrumbs-separator emotion-5"
      >
        /
      </li>
      <li
        class="MuiBreadcrumbs-li"
      >
        <a
          class=""
          data-cy="breadcrumb-item:SUB-LINK"
          href="/sub-link"
        >
          <span
            class="MuiTypography-root MuiTypography-caption emotion-3 emotion-4"
          >
            SUB-LINK
          </span>
        </a>
      </li>
      <li
        aria-hidden="true"
        class="MuiBreadcrumbs-separator emotion-5"
      >
        /
      </li>
      <li
        class="MuiBreadcrumbs-li"
      >
        <span
          class="MuiTypography-root MuiTypography-caption emotion-3 emotion-10"
          data-cy="Current Page Breadcrumb:CURRENT-PAGE"
        >
          CURRENT-PAGE
        </span>
      </li>
    </ol>
  </nav>
</div>
`;
